### PR TITLE
Introduce the torch.distributed.ops module

### PR DESCRIPTION
**Pull request by cbalioglu**
Fixes #58005

This PR introduces a new `torch.distributed.ops` module that replaces the `torch.distributed.nn.functional` module. Instead of exposing the low-level collective communication primitives such as `all_reduce` or `broadcast`, the new module offers a high-level API that offers functions such as `sum`, `prod`, or `copy`. The main motivation for this change was that exposing CC primitives as autograd-aware functions was an inherently flawed approach. In particular computing the gradients for different CC operations involves more than simply calling the specified CC primitive. By exposing high-level functions we hide away which CC primitives are internally used and avoid giving the users a false sense of what is being executed.

### CODE REVIEW

**Comment by ppwwyyxx** [comment]
>  In particular computing the gradients for different CC operations involves more than simply calling the specified CC primitive. By exposing high-level functions we hide away which CC primitives are internally used and avoid giving the users a false sense of what is being executed.

I personally felt that creating new API names (from all_reduce to sum/prod/...) is not necessary.

What primitive gets executed under the hood is just an implementation detail that most if not all users shouldn't care anyway. 
As a user, `dist.nn.all_reduce` is just a mathematical operation to me and when calling it I would not assume how it is implemented. Even if I do have some false sense in my mind about its implementation, what harm can it do to me as long as this function does the right thing?

**Review by mrshenli — COMMENTED** [review]
@ppwwyyxx's makes sense to me. I can see both pros and cons of having `torch.distributed.ops`. 

* On the upside, it's not abusing the same API, and we can clearly distinguish that with non-autograd `all_reduce`. Users will never shoot themselves in the foot. 
* On the downside, users will need to learn a new set of APIs, which might be an overkill? 

Shall we do a team discussion on this design?

**Comment by cbalioglu** [comment]
> Shall we do a team discussion on this design?

Makes sense. I will reserve next Thursday's (08/26) weekly meeting.
